### PR TITLE
Optimize FlatVector<StringView>::copy()

### PR DESCRIPTION
**Pull request by mbasmanova**
Summary:
FlatVector<StringView>::copy goes through different code paths depending on
whether source and destination vectors use the same memory pool or not. Same
pool path is fast. Different pools path is slow. The slow path is taken
whenever we copy strings from a vector produced by one operator into a vector
produced by another operator. For example, during array_constructor function
evaluation in FilterProject over a vector produced by TableScan.

Optimize slow
path by (1) using DecodedVector; (2) calculating total size of strings and
allocated string buffer in one go.

Differential Revision: D50867877

### CODE REVIEW

File: velox/vector/FlatVector.cpp [hunk 1]
```diff
@@ -281,14 +281,44 @@ void FlatVector<StringView>::copy(
...
   } else {
+    DecodedVector decoded(*source);
+    uint64_t* rawNulls = const_cast<uint64_t*>(BaseVector::rawNulls_);
+    if (decoded.mayHaveNulls()) {
+      rawNulls = BaseVector::mutableRawNulls();
+    }
+
+    size_t totalBytes = 0;
     rows.applyToSelected([&](vector_size_t row) {
       auto sourceRow = toSourceRow ? toSourceRow[row] : row;
```
Comment (xiaoxmeng) [inline velox/vector/FlatVector.cpp:292]:
```
nit: const auto sourceRow
```

File: velox/vector/FlatVector.cpp [hunk 2]
```diff
@@ -281,14 +281,44 @@ void FlatVector<StringView>::copy(
     copyValuesAndNulls(source, rows, toSourceRow);
     acquireSharedStringBuffers(source);
   } else {
+    DecodedVector decoded(*source);
```
Comment (xiaoxmeng) [inline velox/vector/FlatVector.cpp:284]: I am curious why using decoded vector here help performance? Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova nice optimization. Thanks!

File: velox/vector/FlatVector.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/vector/FlatVector.cpp:284]: It helps avoid virtual function calls to check isNullAt(sourceRow) and to get valueAt(sourceRow).

**Pull request closed without merging**